BaseClasses.py

```diff
@@ -681,19 +681,57 @@ def can_reach_region(self, spot: str, player: int) -> bool:
         return self.multiworld.get_region(spot, player).can_reach(self)
 
     def sweep_for_events(self, key_only: bool = False, locations: Optional[Iterable[Location]] = None) -> None:
-        if locations is None:
-            locations = self.multiworld.get_filled_locations()
-        reachable_events = True
+        events_per_player: Dict[int, Set[Location]]
+
         # since the loop has a good chance to run more than once, only filter the events once
-        locations = {location for location in locations if location.advancement and location not in self.events and
-                     not key_only or getattr(location.item, "locked_dungeon_item", False)}
-        while reachable_events:
-            reachable_events = {location for location in locations if location.can_reach(self)}
-            locations -= reachable_events
+        if key_only:
+            def event_filter(location: Location):
+                return location.item is not None and getattr(location.item, "locked_dungeon_item", False)
+        else:
+            def event_filter(location: Location):
+                return (location.advancement and location not in self.events
+                        or getattr(location.item, "locked_dungeon_item", False))
+        if locations is None:
+            # `self.multiworld.get_filled_locations(player)` is avoided because it first iterates into a list and also
+            # because `location.advancement` in `event_filter` also checks for `location.item is not None`.
+            events_per_player = {player: set(filter(event_filter, locations_dict.values()))
+                                 for player, locations_dict in self.multiworld.regions.location_cache.items()}
+        else:
+            # Can't iterate self.multiworld.player_ids because it doesn't include the extra player IDs used for item
+            # links, so iterate the keys of the location_cache instead.
+            events_per_player = {player: set() for player in self.multiworld.regions.location_cache.keys()}
+            for location in locations:
+                if event_filter(location):
+                    events_per_player[location.player].add(location)
+
+        # Remove empty sets to reduce the number of iterations required to iterate through `events_per_player.items()`.
+        for player, events in list(events_per_player.items()):
```

Comment:
is this actually faster than just leaving them be? iterating over an empty set shouldn't cost anything.

Reply:
It's to do with the length of the dictionary itself rather than the sets because the while loop iterates `events_per_player.items()` many times, so reducing the length of `events_per_player` should make it faster to iterate.